File: vertex_api.py
from flask import Flask, request, jsonify
from google.cloud import aiplatform
from jsonschema import validate, ValidationError
import os
import logging

app = Flask(__name__)
logging.basicConfig(level=logging.INFO)

# Configuration
PROJECT_ID = os.environ.get("PROJECT_ID")  # GCP project ID
REGION = os.environ.get("REGION")          # GCP region (e.g., "us-central1")
ENDPOINT_ID = os.environ.get("ENDPOINT_ID") # Vertex AI Endpoint ID

# Initialize Vertex AI client
aiplatform.init(project=PROJECT_ID, location=REGION)
endpoint_client = aiplatform.Endpoint(endpoint_name=ENDPOINT_ID)

# Define schema for input validation
schema = {
    "type": "object",
    "properties": {
        "instances": {
            "type": "array",
            "items": {
                "type": "array",
                "items": {"type": "number"}
            }
        }
    },
    "required": ["instances"]
}

@app.route('/predict', methods=['POST'])
def predict():
    try:
        request_json = request.get_json()
        if not request_json:
            return jsonify({"error": "Request must be in JSON format", "code": "INVALID_FORMAT"}), 400

        try:
            validate(instance=request_json, schema=schema)
        except ValidationError as ve:
            return jsonify({"error": "Invalid input format", "code": "INVALID_INPUT", "details": str(ve)}), 400

        instances = request_json["instances"]
        try:
            prediction = endpoint_client.predict(instances=instances)
            return jsonify(prediction.predictions)
        except GoogleAPIError as e:
            logging.error(f"Vertex AI error: {e}")
            return jsonify({
                "error": "Vertex AI prediction failed",
                "code": "VERTEX_AI_ERROR",
                "details": str(e)
            }), 500
    except Exception as e:
        logging.error(f"Unexpected error: {e}")
        return jsonify({
            "error": "Internal server error",
            "code": "INTERNAL_ERROR",
            "details": "An unexpected error occurred. Please try again later."
        }), 500

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=int(os.environ.get("PORT", 8080))) 

File: vertex_api.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"google.golang.org/api/aiplatform/v1"
	"google.golang.org/api/option"
)

// VertexAIClient is a wrapper for the Vertex AI client
type VertexAIClient struct {
	client *aiplatform.Service
}

// NewVertexAIClient creates a new Vertex AI client
func NewVertexAIClient(projectID, location, endpointID string) (*VertexAIClient, error) {
	ctx := context.Background()
	client, err := aiplatform.NewService(ctx, option.WithCredentialsFile(os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")))
	if err != nil {
		return nil, fmt.Errorf("error creating Vertex AI client: %v", err)
	}
	return &VertexAIClient{client: client}, nil
}

// Predict sends a prediction request to Vertex AI
func (c *VertexAIClient) Predict(instances [][]float64) ([]interface{}, error) {
	req := &aiplatform.GoogleCloudAiplatformV1PredictRequest{
		Instances: instances,
	}
	resp, err := c.client.Projects.Locations.Endpoints.Predict("projects/"+os.Getenv("PROJECT_ID")+"/locations/"+os.Getenv("REGION")+"/endpoints/"+os.Getenv("ENDPOINT_ID"), req).Do()
	if err != nil {
		return nil, fmt.Errorf("error making prediction: %v", err)
	}
	return resp.Predictions, nil
}

// InputSchema defines the structure for input validation
type InputSchema struct {
	Instances [][]float64 `json:"instances"`
}

func main() {
	http.HandleFunc("/predict", func(w http.ResponseWriter, r *http.Request) {
		var input InputSchema
		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
			http.Error(w, fmt.Sprintf("{\"error\": \"Request must be in JSON format\", \"code\": \"INVALID_FORMAT\"}"), http.StatusBadRequest)
			return
		}

		if len(input.Instances) == 0 {
			http.Error(w, fmt.Sprintf("{\"error\": \"JSON must contain 'instances' key\", \"code\": \"MISSING_KEY\"}"), http.StatusBadRequest)
			return
		}

		client, err := NewVertexAIClient(os.Getenv("PROJECT_ID"), os.Getenv("REGION"), os.Getenv("ENDPOINT_ID"))
		if err != nil {
			http.Error(w, fmt.Sprintf("{\"error\": \"Internal server error\", \"code\": \"INTERNAL_ERROR\", \"details\": \"%s\"}", err.Error()), http.StatusInternalServerError)
			return
		}

		predictions, err := client.Predict(input.Instances)
		if err != nil {
			log.Printf("Vertex AI error: %v", err)
			http.Error(w, fmt.Sprintf("{\"error\": \"Vertex AI prediction failed\", \"code\": \"VERTEX_AI_ERROR\", \"details\": \"%s\"}", err.Error()), http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(predictions)
	})

	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}
	log.Printf("Starting server on port %s", port)
	log.Fatal(http.ListenAndServe(":"+port, nil))
}
